### DISCUSSION TITLE
Mastodon Comments Plugin

### DISCUSSION THREAD

**Discussion by happybeing** [GetPublii/Publii#1857] [Ideas]
I've just set up my Publii blog and the only thing that it needs now is a way for people to comment - that doesn't really on a commercial service. 

For me the best option would be Mastodon / ActivityPub. I believe this exists for WordPress so I'm hoping this would also be feasible for Publii.

I realise it would have to work differently so I imagine that when you create a post you would have a way for it to be posted to your Mastodon account and that Publii would store the toot URL with metadata for the post. 

A comment plugin could use the URL to let a commenter create a reply from their own Mastodon account, which would be displayed on the Publii blog as a comment (by fetching all the replies to the toot and showing them as comments).

**Comment by candidexmedia** [comment]
This is a great idea! I've been casually collecting different implementations of Mastodon comments inside of a static site, and can share what I found below:

- https://danielpecos.com/2022/12/25/mastodon-as-comment-system-for-your-static-blog/ : Not bad, but doesn’t look great on mobile
    - ⇒ Easy web component: https://github.com/dpecos/mastodon-comments (simplest approach)
- Precursor of the above: https://carlschwan.eu/2020/12/29/adding-comments-to-your-static-blog-with-mastodon/
    - Hacker News comments: https://news.ycombinator.com/item?id=25570268
- https://grahammacphee.com/writing/mastodon-blog-comments: looks good
    - Code: https://gist.github.com/gmph/02a44fc28f7da91da05462f7046ab19f
- https://cassidyjames.com/blog/fediverse-blog-comments-mastodon/ (no code yet, looks really nice)
- https://jszym.com/blog/mastodon_blog_comments/ (not the cutest looking, but good writeup)

Haven't tested them in Publii yet, but each one requires some degree of getting your hands dirty in code, or self-hosting something. If you try some out, please report back with your findings! :) 

Keep in mind though: people replying to your post may not necessarily know / consent to having their avatars, names, and comments appear on your site. Here are some articles discussing some of the challenges and pitfalls of Mastodon site comments / webmentions / indieweb:
- https://shkspr.mobi/blog/2022/12/the-ethics-of-syndicating-comments-using-webmentions/
- https://sebastiangreger.net/2018/05/indieweb-privacy-challenge-webmentions-backfeeds-gdpr/
- https://robertkingett.com/posts/6408/
- https://brainbaking.com/post/2023/05/why-i-retired-my-webmention-server/
- https://darekkay.com/blog/static-site-comments/

You could also argue that "[every mastodon host out there [is] a website](https://mastodon.nu/@rkaj/111912567225865834)" in the same way that yours is, but it's something to keep in mind. Perhaps making a mention of it in your profile?

Otherwise, I do want to note that there are non-commercial, open source comment services — one of which even has a Publii plugin:
- Cusdis: has a [Publii plugin](https://marketplace.getpublii.com/plugins/cusdis-comments/), and can be used with the [hosted version](https://cusdis.com/) or [self-hosted](https://cusdis.com/doc#/self-host/vercel)
- [Waline](https://waline.js.org/en/): which can be [self-hosted](https://waline.js.org/en/guide/get-started/) pretty much for free
- There's, of course, Commento and its offshoots, but I think the main project [hasn't been maintained in a while,](https://gitlab.com/commento/commento/-/issues/414) last I checked

**Reply by candidexmedia** [reply]
Oooh, some more fun stuff regarding Mastodon / Activity Pub integrations with static sites like Publii:

- You can verify your Publii site so that it appears in plugins like [Streetpass](https://streetpass.social/): see https://github.com/GetPublii/Publii/discussions/1485#discussioncomment-9157568

- You can add an author attribution byline, so that whenever an article of yours is shared in Mastodon, people can find your Mastodon profile (even if someone else posted the link): see https://github.com/GetPublii/Publii/discussions/1485#discussioncomment-10923402

- You can automate the posting of your articles to Mastodon using your Publii RSS feed with MastoFeed: https://mastofeed.org/

- Mastodon share button: https://mastodonshare.com/about

- Your site could serve as its own Fediverse account, receiving @ mentions as webmentions using [Bridgyfed](https://fed.brid.gy/). I was able to try this out and get it working (my site can be followed on the fedi as [@candide.media@candide.media](https://mastodon.social/@candide.media@candide.media)), but I haven't tested the conversion of posts to microformats yet so that they automatically appear on the feed.

**Comment by Nevarro** [comment]
@happybeing 

I just thought about a feature like that and my ideas for an implementation match yours 1:1. So I think it would be great to have a plugin for Wordpress, that uses Mastodon for comments. I think Publii's open source idea fits perfectly to Mastodon's ideals.

**Comment by nozense** [comment]
Activitypub comments or/and site as fediverse "user" would be great. Id gladly pay for a plugin.

atm there is no viable way to enable comments without paying or selfhost a platform. This is not practical for me and i'm surely not alone. And why not make the fediverse a bit more lively and publii better :)

**Reply by candidexmedia** [reply]
> atm there is no viable way to enable comments without paying or selfhost a platform.

Comments in general, or just fediverse comments?

If it's comments in general: [GraphComment](https://marketplace.getpublii.com/plugins/graphcomment/) is [free](https://www.graphcomment.com/en/products/graphcomment/pricing).

**Reply by nozense** [reply]
Thank you! I had actually overlooked that option. 
But yes, primarily I'm after something that does not require me to sign up at a private company. I try to keep as much as possible open source and privacy friendly (not saying anything about GraphComments).
And also yes, I like the fediverse and hope it's the future - so specifically a ActivityPub solution would make my day :)
